ICDS scheduler modules

proteusvacuum opened the pull request:
# ICDS Scheduler Module

This adds the functionality to create scheduler modules in HQ.
- [Spec](https://docs.google.com/document/d/18Sw28SJ325G-LkoEfGClcKGyJPGnSk-jDaqkNw6FS1g/edit#heading=h.b0sotnmxj18r)
- [Original mockup + technical details](https://confluence.dimagi.com/display/ccinternal/Scheduler+Module)
- [Final Mockup + Technical Details](https://confluence.dimagi.com/display/ccinternal/Scheduler+Module+2)

![screen shot 2015-08-26 at 6 07 06 pm](https://cloud.githubusercontent.com/assets/146896/9507575/4bfdb326-4c1d-11e5-8fcf-11c1120b58a5.png)

![screen shot 2015-08-26 at 6 07 33 pm](https://cloud.githubusercontent.com/assets/146896/9507596/5d9a648a-4c1d-11e5-81c3-f7e478a19474.png)

The spec is a little hard to parse without context, so...
## Schedule Phase

The primary unit of scheduling. You can transition between phases which sticks you on a different 'track'. For example "Pre Birth" or "After Birth". The phase is linked with an 'anchor date' which is a case property. A phase has a number of forms, which has a number of visits.
## Form relevancy window:

You can set properties for when forms are relevant. Outside of this relevancy window, forms are hidden.
## Visit relevancy window:

Visits have relevancy window. If a visit occurs outside of all of the visit windows it is deemed to be 'unscheduled' (see 'form properties', below). 
When 'allow unscheduled visits' is checked, these forms are always shown, otherwise they are only shown when the current date falls within the relevancy window.
## Repeat Visits:

The last visit can be set as a repeat visit, which will become 'due' with the properties you set (set the dropdown for the last visit to be 'after the last visit'). If there is a repeat visit, you can't add any more fixed visits after it (even though I think in theory you should be able to with the queries we have).
## Detail variables:

For scheduler modules there are `next_due` and `is_late` in the case details screen by adding `schedule:is_late` and `schedule:next_due`. You'll need to add at least one of these for the scheduler module to work right now.
The next_due variable returns the date of the next due visit. (Returns the date of the heatdeath of the universe if there are no scheduled visits (not really))
The is_late flag is true if this visit is... late.

For debugging (and calculation) purposes you also have:
- schedule:last_visit_number_{form_id} : which returns the last visit for the selected case for the form with form_id
- schedule:next_{form_id}: the next visit date for form_id (returns an int)
- schedule:anchor_{form_id}: the anchor date that this form thinks it should be using
## Form properties:

In the form you now have access to:
- `/data/unscheduled_visit` : this is set to `true()` if the visit is deemed to be unscheduled (i.e. outside of the visit relevancy window). This is not saved to the case, but is sent back to the server with the form.
- `data/next_visit_number` which returns the next scheduled visit number. This is saved to the case as `last_visit_num_{form_id}` when this is a scheduled visit, and the form has been submitted.

Adding these properties required me to add a reference to the fixture in each form, so you'll also be able to get raw schedule properties with `instance('schedule:m{num}:p{phase_num}:f{form_num}')`... although, this isn't that user friendly. If this is something that you see as being useful, I could change the instance id to be something nicer (e.g. just 'schedule') in the form.
## Case properties:
- `last_visit_num_{form_id}`: gives you the number of the last scheduled visit that was undertaken (i.e. if you missed visit 2, then did visit 3, this would show '3', if you then carried out an unscheduled visit, this would still be '3')
- `last_visit_date_{form_id}`: gives the date of when `last_visit_num_{form_id}` was carried out.
- `current_schedule_phase` is updated to the current phase.

@snopoke @ctsims 
buddies: @esoergel @emord (sorry!)
## TODO
- [x] Use arbitrary form abbreviations
- [x] UI to create and re-order phases on module page
  - [x] Allow user to select from these phases on the visit scheduler UI
- [x] Change next_visit_number -> current_visit_number in form 
- [x] Add option to always show a form (i.e. not part of scheduler)
- [x] Add global `next_visit_date` to each form.

proteusvacuum committed:
Multiple phases per module

proteusvacuum committed:
Multiple forms per phase fixtures

proteusvacuum committed:
Refactor

proteusvacuum committed:
Form Filtering

proteusvacuum committed:
get_or_create_schedule_phase method

proteusvacuum committed:
WIP: get UI working with new phase model

proteusvacuum committed:
Don't nest forms in SchedulePhases

Reference an ID instead!

proteusvacuum committed:
Updates to view

proteusvacuum committed:
current_schedule_phase setvalue

proteusvacuum committed:
last visit number and last visit date tests

proteusvacuum committed:
Bind instead of setvalue for case properties

proteusvacuum committed:
Use casedb for current_schedule_phase for form filtering

proteusvacuum committed:
Transition and termination conditions

proteusvacuum committed:
no staticmethod for detail variables

proteusvacuum committed:
Add extra detail variables

Add detail variables so we can reference them in xpath predicates

proteusvacuum committed:
Use case action property in form filter

Advanced modules don't use the /case_id property, instead each load
action is given a tag. This commit makes use of this tag

proteusvacuum committed:
Make filter conditions work

Especially if there are no prior visit

proteusvacuum committed:
Force 'expires' as a property on fixture

proteusvacuum committed:
Acutally update case properties

Advanced Forms load cases differently

proteusvacuum committed:
fixup: docstring fixup

proteusvacuum committed:
Add last_visit_date update caseblock element

dimagimon commented:
There are 6 errors:

* corehq/apps/app_manager/models.py, line 2017 - E261 at least two spaces before inline comment
* corehq/apps/app_manager/models.py, line 2236 - E301 expected 1 blank line, found 0
* corehq/apps/app_manager/models.py, line 2571 - E261 at least two spaces before inline comment
* corehq/apps/app_manager/tests/test_schedule.py, line 301 - E501 line too long (116 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 322 - E501 line too long (117 > 115 characters)
* corehq/apps/app_manager/views.py, line 407 - E302 expected 2 blank lines, found 1

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -2006,6 +2011,23 @@ def all_other_forms_require_a_case(self):
+        if isinstance(self._parent, ModuleBase):
+            return self._parent
+        else: # Forms can be nested within SchedulePhases
+            return self._parent._parent
```
Is this still needed?

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -2086,7 +2108,7 @@ def extended_build_validation(self, error_meta, xml_valid, validate_module=True)
 
         module = self.get_module()
-        if module.has_schedule and not (self.schedule and self.schedule.anchor):
+        if module.has_schedule and not self.schedule and getattr(self.schedule, 'anchor', False):
```
Won't `getattr(self.schedule, 'anchor', False)` always be False since the anchor is now in SchedulePhase

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -2178,6 +2200,88 @@ def update_app_case_meta(self, app_case_meta):
+    """
+    A reference to a form in a schedule phase.
+    """
+    form_id = FormIdProperty("modules[*].schedule_phases[*].form_id")
```
I think this should be `modules[*].schedule_phases[*].forms[*].form_id`

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -2178,6 +2200,88 @@ def update_app_case_meta(self, app_case_meta):
+    forms: 			The forms that are to be filled out within this phase
+    """
+    anchor = StringProperty()
+    forms = SchemaListProperty(SchedulePhaseForm)
```
Is it not possible to just make a `FormIdListProperty` that returns a `StringListProperty`

snopoke replied:
Unless you need the IndexedSchema functionality

proteusvacuum replied:
Yeah, I use the IndexedSchema functionality in `get_phase_form_index` when making the fixture

snopoke commented on corehq/apps/app_manager/static/app_manager/js/visit-scheduler.js:
```diff
@@ -79,11 +80,27 @@ var VisitScheduler = (function () {
+                include: [
+                    'anchor',
+                    'transition_condition',
+                    'termination_condition'
```
Aren't these in the FormSchedule?

proteusvacuum replied:
yup! for a while I thought they were on the phase. forgot to remove this.

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1265,86 +1267,43 @@ def details(self):
 
-    @staticmethod
-    def detail_variables(module, detail, detail_column_infos):
+    def _schedule_detail_variables(self, module, detail, detail_column_infos):
```
I think this can be a staticmethod, no?

snopoke commented:
First pass, looking good.

proteusvacuum committed:
preliminary PR feedback

dimagimon commented on corehq/apps/app_manager/tests/test_schedule.py:
```diff
@@ -0,0 +1,371 @@
+        )
+        self.assertXmlPartialEqual(
+            current_schedule_phase_partial.format(value=value, xmlns=self.xmlns),
+            (xform_1.model_node.find('./bind[@nodeset="/data/case_case_clinic/case/update/current_schedule_phase"]')
```
E501 line too long (116 > 115 characters)

dimagimon commented on corehq/apps/app_manager/tests/test_schedule.py:
```diff
@@ -0,0 +1,371 @@
+        self.form_2.add_stuff_to_xform(xform_2)
+        self.assertXmlPartialEqual(
+            current_schedule_phase_partial.format(value=value, xmlns=self.xmlns),
+            (xform_2.model_node.find('./bind[@nodeset="/data/case_load_clinic0/case/update/current_schedule_phase"]')
```
E501 line too long (117 > 115 characters)

proteusvacuum committed:
Delete form from phase when form is deleted

proteusvacuum committed:
Relvancy window in UI

proteusvacuum committed:
Add relevancy window and allow_unscheduled to UI

proteusvacuum committed:
Add new visit properties to fixture

proteusvacuum committed:
Add extra properties to schedule fixture

proteusvacuum committed:
New filter condition

dimagimon commented:
There are 6 errors:

* corehq/apps/app_manager/tests/test_schedule.py, line 235 - E501 line too long (173 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 237 - E501 line too long (120 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 240 - E501 line too long (127 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 241 - E501 line too long (145 > 115 characters)
* corehq/apps/app_manager/xpath.py, line 438 - E501 line too long (186 > 115 characters)
* corehq/apps/app_manager/xpath.py, line 439 - E501 line too long (128 > 115 characters)

proteusvacuum committed:
Fixup: make 'add new visit' button work again

dimagimon commented:
There are 6 errors:

* corehq/apps/app_manager/tests/test_schedule.py, line 235 - E501 line too long (173 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 237 - E501 line too long (120 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 240 - E501 line too long (127 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 241 - E501 line too long (145 > 115 characters)
* corehq/apps/app_manager/xpath.py, line 438 - E501 line too long (186 > 115 characters)
* corehq/apps/app_manager/xpath.py, line 439 - E501 line too long (128 > 115 characters)

proteusvacuum committed:
Make math work

proteusvacuum committed:
unscheduled_visits -> allow_unscheduled

proteusvacuum committed:
Make within_window form filter query work

dimagimon commented:
There are 6 errors:

* corehq/apps/app_manager/tests/test_schedule.py, line 235 - E501 line too long (173 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 237 - E501 line too long (120 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 240 - E501 line too long (127 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 241 - E501 line too long (158 > 115 characters)
* corehq/apps/app_manager/xpath.py, line 438 - E501 line too long (187 > 115 characters)
* corehq/apps/app_manager/xpath.py, line 439 - E501 line too long (141 > 115 characters)

dimagimon commented:
There are 6 errors:

* corehq/apps/app_manager/tests/test_schedule.py, line 235 - E501 line too long (174 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 237 - E501 line too long (120 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 240 - E501 line too long (127 > 115 characters)
* corehq/apps/app_manager/tests/test_schedule.py, line 241 - E501 line too long (158 > 115 characters)
* corehq/apps/app_manager/xpath.py, line 438 - E501 line too long (187 > 115 characters)
* corehq/apps/app_manager/xpath.py, line 439 - E501 line too long (141 > 115 characters)

proteusvacuum committed:
next_ queries

proteusvacuum committed:
Get detail variables to show up on detail screen

proteusvacuum committed:
Add next_visit_number and unscheduled_visit to form

- also 'coalesce' the next_ variables so we don't get NaN's

dimagimon commented on corehq/apps/app_manager/const.py:
```diff
@@ -24,6 +24,10 @@
+SCHEDULE_TERMINATED = '-1'
+SCHEDULE_NEXT_VISIT_NUMBER = 'next_visit_number'
+SCHEDULE_UNSCHEDULED_VISIT = 'unscheduled_visit'
+SCHEDULE_MAX_DATE = (2**31) - 1
```
E226 missing whitespace around arithmetic operator

proteusvacuum committed:
Don't set last_visit_date in unscheduled visits

proteusvacuum committed:
Don't fail when no visits defined

proteusvacuum committed:
add last_visit_date variable

proteusvacuum committed:
Allow form relevancy to start at 0

proteusvacuum committed:
fix case sharing issue

proteusvacuum committed:
Show next visits even if we are not within window

proteusvacuum committed:
Small wording changes

proteusvacuum committed:
Fix failing test

snopoke commented on corehq/apps/app_manager/tests/test_schedule.py:
```diff
@@ -0,0 +1,381 @@
+                "today() &lt;= (date({case}/last_visit_date_{form_id}) + int(@increment) + int(@expires))), "
+                "today() &gt;= (date({anchor}) + int(@due) + int(@starts)) and "
+                "(@expires = '' or today() &lt;= (date({anchor}) + int(@due) + int(@expires))))"
+                "]) &gt; 0)"     # End count
```
This is so horrible but can't think of a better way to test it.

proteusvacuum replied:
hah! I'm going to break it out into separate variables so it reads better, and matches the `SchedulerXPath` properties that these get derived from.
eg: 
`within_window = "today() >= (date({anchor}) + int({schedule}/@starts)"..."` 
etc. 
I think the one in the test fixture file is actually more unwieldy since you can't add linebreaks, and it is just a huge blob of xpath that I am testing.

proteusvacuum committed:
Dimagimon + test refactor

dimagimon commented on corehq/apps/app_manager/tests/test_schedule.py:
```diff
@@ -0,0 +1,408 @@
+        """ Current Schedule Phase is set depending on transition and termination conditions """
+        self._fetch_sources()
+
+        current_schedule_phase_partial ="""
```
E225 missing whitespace around operator

proteusvacuum committed:
Disable form filtering for schedule modules

proteusvacuum committed:
Merge branch 'master' into icds/scheduler

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into icds/scheduler

esoergel commented on corehq/apps/app_manager/tests/test_schedule.py:
```diff
@@ -0,0 +1,408 @@
+    def setUp(self):
+        self.is_usercase_in_use_patch = patch('corehq.apps.app_manager.models.is_usercase_in_use')
+        self.is_usercase_in_use_mock = self.is_usercase_in_use_patch.start()
+        self.is_usercase_in_use_mock.return_value = True
```
Minor point, but I believe if you decorate the class with this patch, it'll be applied to each method that begins with `"test"`.  I'm not sure about specifying a return value in the decorator though, but I think something like this would work:
```python
@patch('corehq.apps.app_manager.models.is_usercase_in_use', lambda: True)
```

esoergel commented on corehq/apps/app_manager/tests/test_schedule.py:
```diff
@@ -0,0 +1,408 @@
+        self.assertXmlPartialEqual(expected_fixture, suite, './fixture[@id="schedule:m2:p1:f0"]')
+        self.assertXmlHasXpath(suite, './fixture[@id="schedule:m1:p1:f0"]')
+
+    def test_form_filtering(self):
```
Other reviewers: this is a useful test to read to understand what this is doing.

esoergel commented:
I read some of this, and it looks good - nothing really constructive to add yet, although I'll try to read more later.  Note that I'll be offline for a while starting tomorrow afternoon, so I won't be able to follow up on this once you're ready to merge.

emord commented on corehq/apps/app_manager/templates/app_manager/form_view_advanced.html:
```diff
@@ -32,7 +32,8 @@
         save_url: "{% url "corehq.apps.app_manager.views.edit_visit_schedule" app.domain app.id module.id nav_form.id %}",
-        schedule: {{ form.schedule|JSON }}
+    	schedule: {{ form.schedule|JSON }},
+    	phase: {{ schedule_phase|JSON }}
```
Mixed tabs and spaces?

proteusvacuum committed:
Merge branch 'master' into icds/scheduler

proteusvacuum committed:
Merge branch 'master' into icds/scheduler

proteusvacuum committed:
Use a user-specified form abbreviation

proteusvacuum committed:
Allow non-ascii characters in abbreviation name

proteusvacuum committed:
Convert abbreviation to snake_case

dimagimon commented on corehq/apps/app_manager/tests/test_schedule.py:
```diff
@@ -0,0 +1,422 @@
+
+
+
+    def tearDown(self):
```
E303 too many blank lines (3)

proteusvacuum committed:
UI to manage schedule phases

proteusvacuum committed:
Update phases with a list of anchors

proteusvacuum committed:
Allow custom regex on snake case

proteusvacuum committed:
Update schedule phase anchor properties from UI

proteusvacuum committed:
select form phase from dropdown of pre-created phases

proteusvacuum committed:
Small fixes

Use change_anchor method to catch blank anchors

Don't break if you select no anchor

Don't fail when not all forms require a case

proteusvacuum committed:
hide form filtering, not case list filtering

proteusvacuum committed:
next_visit_number -> current_visit_number

proteusvacuum committed:
ko remove not destroy

proteusvacuum committed:
No duplicate phases

proteusvacuum committed:
Allow forms to not take part in scheduling

proteusvacuum committed:
Only hide form filtering for advance modules

dimagimon commented on corehq/apps/app_manager/views.py:
```diff
@@ -528,8 +550,10 @@ def get_form_view_context_and_template(request, form, langs, is_user_registratio
-        'allow_form_workflow': not isinstance(form, CareplanForm),
+        'allow_form_filtering': not isinstance(form, CareplanForm) and not (
+            (isinstance(module, AdvancedModule) and module.has_schedule)),
+        'allow_form_workflow': not isinstance(form, CareplanForm)  and not (
```
E272 multiple spaces before keyword

proteusvacuum committed:
Make enable visit scheduler button work again

proteusvacuum committed:
Dont fail on old apps with advanced modules

Apps which had forms before the scheduler module existed were breaking
because there was no scheduler property.
There was also a catch 22 where you couldn't enable scheduling before
you had case forms, but forms created before having the schedule would
break things because they didn't have the schedule property

proteusvacuum committed:
Merge branch 'master' into icds/scheduler

dimagimon commented on corehq/apps/app_manager/models.py:
```diff
@@ -2493,6 +2612,59 @@ def uses_usercase(self):
         return self._uses_case_type(USERCASE_TYPE)
 
 
+    @property
```
E303 too many blank lines (2)

dimagimon commented on corehq/apps/app_manager/xpath.py:
```diff
@@ -1,6 +1,13 @@
+    SCHEDULE_TERMINATED,
+    SCHEDULE_MAX_DATE,
+)
+from corehq.apps.app_manager.exceptions import LocationXpathValidationError, ScheduleError, CaseXPathValidationError
```
E501 line too long (116 > 115 characters)

proteusvacuum committed:
:gift: for dimagimon and fix tests

proteusvacuum committed:
Add next_visit properties to each form

proteusvacuum committed:
Add test for next_visit_date

proteusvacuum committed:
Add label to empty form logic

http://manage.dimagi.com/default.asp?180063#1001953

proteusvacuum committed:
Moving forms between advanced modules

proteusvacuum committed:
Front end validation

proteusvacuum committed:
Add change triggers when new visit is added

proteusvacuum commented:
@snopoke I think I'm close to being "done" here. I think I fixed all of the HQ QA bugs that Clark found. He's working on verifying all the logic on mobile too.
Was hoping to get this merged and onto India so that they can test it in their app before the demo.
@orangejenny, you are my new code buddy, could you give it a pass too if you have a moment?

@fyniac - ping

sheelio commented:
@proteusvacuum Farid, just checking we can get this deployed to India tonight?

fyniac commented:
@snopoke @orangejenny @proteusvacuum  can you please prioritize today?  Want to make sure this gets on the India server so Sheel & gang can use it in their app

orangejenny commented:
yep, will review now (fyi @proteusvacuum there are merge conflicts at the moment)

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -687,9 +706,12 @@ def get_form(cls, form_unique_id, and_app=False):
-    @property
-    def schedule_form_id(self):
-        return self.unique_id[:6]
+    def _pre_delete_hook(self):
```
curious why this is marked private, looks like it's used outside this class

proteusvacuum replied:
good point. will change.

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -1983,6 +2005,19 @@ def wrap(cls, data):
+    def _pre_delete_hook(self):
+        try:
+            self.disable_schedule()
+        except (ScheduleError, TypeError, AttributeError):
```
no logging / notification? when would these exceptions be thrown / why doesn't it matter if they are?

proteusvacuum replied:
Yeah, I should probably add a log here. These exceptions are only thrown if something has gone out of whack (e.g. the form thinks it is in a phase, but the phase disagrees). `AttributeError` gets thrown if the form was created before this PR, i.e. if it doesn't have a `schedule` property. 
Basically, I didn't think that anything bad was happening if these exceptions are raised which is why I didn't handle them past just skipping over it.

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -2226,6 +2270,88 @@ def update_app_case_meta(self, app_case_meta):
+
+    @property
+    def id(self):
+        """ A Schedule Phase is 1-indexed """
```
just curious: why is this?

proteusvacuum replied:
When I started building this I thought that phase 0 would be referenced when the case had not entered into a schedule. This ended up being unnecessary since I check current_schedule_phase = '' to be the 0th phase, but I never changed this back. I'm going to leave it this way for now since it matches the numbering on the edit phases page (and 0-based indexing is weird when it is user-facing)

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into icds/scheduler

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -2226,6 +2270,88 @@ def update_app_case_meta(self, app_case_meta):
+        module = self.get_module()
+        return (module.get_form_by_unique_id(form.form_id) for form in self._get_forms())
+
+    def get_form(self, desired_form):
```
not following the purpose of this: you give it a form object and it gives you back...a different representation of the same form?

proteusvacuum replied:
yeah... it's a bit strange, but this returns the form within the phase (with it's `id` within the phase that form is in)

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -2226,6 +2270,88 @@ def update_app_case_meta(self, app_case_meta):
+
+        self.anchor = new_anchor
+
+        if self.get_module().phase_anchors.count(new_anchor) > 1:
```
so if this happens, self has already had its anchor updated...is that going to be a problem?

snopoke replied:
I think the exception will prevent the app from being saved so the changes won't get persisted.

snopoke replied:
Would seem better to do the check prior to making the change though.

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -2492,6 +2619,58 @@ def uses_usercase(self):
+
+        if phase is None:
+            self.schedule_phases.append(SchedulePhase(anchor=anchor))
+            # TODO: is there a better way of doing this?
```
I guess you could manually call `phase.with_id(index, parent)`

snopoke replied:
Not sure if that's any better though

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -2510,6 +2637,58 @@ def uses_usercase(self):
+        if deleted_phases_with_forms:
+            raise ScheduleError(_("You can't delete phases with anchors "
+                                  "{phase_anchors} because they have forms attached to them").format(
+                                      phase_anchors=(", ").join(deleted_phases_with_forms)))
```
Similar to question above about making changes to the app.

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -2405,22 +2394,34 @@ def menus(self):
+                        elif form.requires_case():
+                            case = SESSION_CASE_ID.case()
+                        else:
+                            case = None
```
Why did this get moved out of the `if` statement?

snopoke replied:
Ah, I see you needed the `case` ref

snopoke commented on corehq/apps/app_manager/tests/data/suite/schedule-entry.xml:
```diff
@@ -6,11 +6,20 @@
+            <last_visit_date_39f0ea function="last_visit_date_39f0ea"/>
+            <next_39f0ea function="if(current_schedule_phase = 2 and $anchor_39f0ea != '' and (today() &gt;= (date($anchor_39f0ea) + int(instance('schedule:m1:p2:f0')/schedule/@starts)) and (instance('schedule:m1:p2:f0')/schedule/@expires = '' or today() &lt;= (date($anchor_39f0ea) + int(instance('schedule:m1:p2:f0')/schedule/@expires)))), date($anchor_39f0ea) + int(coalesce(instance('schedule:m1 ...
+            <next_due function="date(min($next_हिन्दी,$next_a983e9,$next_39f0ea))"/>
+            <is_late function="$next_due &lt; today()"/>
```
yikes!

proteusvacuum replied:
:imp: 
This was the bane of my existence for a while. I should have extracted that statement from here and composed it directly in the test. But... it's done now.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/visit-scheduler.js:
```diff
@@ -3,22 +3,38 @@
+                    .on('click', 'a:not(.header)', self.change)
+                    .on('change', 'input[type="checkbox"]', self.change);
+
+                // https://gist.github.com/mkelly12/424774/#comment-92080
```
picky, but would be nice to add `//textchange doesn't work with live event binding`

orangejenny replied:
oh, i see you just moved this. still would be nice, if you're feeling energetic.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/visit-scheduler.js:
```diff
@@ -39,7 +133,35 @@ var VisitScheduler = (function () {
         };
 
-        self.formSchedule = FormSchedule.wrap(params.schedule, self);
+        self.validate = function(){
```
Is this the first place HQ is doing client-side required fields validation? That's kind of fascinating.

snopoke replied:
App builder does this in a few places.

proteusvacuum replied:
Yeah, couldn't find any good examples, so I made it up. Might be useful to extract this into `saveButton` if this will become a common pattern.

snopoke commented on corehq/apps/app_manager/xform.py:
```diff
@@ -1511,7 +1571,8 @@ def check_case_type(action):
                 if has_schedule:
-                    configure_visit_schedule_updates(update_case_block.update_block)
+                    self.add_casedb()
+                    configure_visit_schedule_updates(update_case_block.update_block, action)
```
Edge case but I think you should pass `session_case_id` into this function rather than getting it from the action since the action may not have the correct value (in the case of submodules).

snopoke commented on corehq/apps/app_manager/xform.py:
```diff
@@ -1404,36 +1407,93 @@ def make_delegation_stub_case_block():
+                nodeset=u'/data/{}'.format(SCHEDULE_UNSCHEDULED_VISIT),
+                calculate=schedule_form_xpath.is_unscheduled_visit,
+            )
+            self.data_node.append(_make_elem(SCHEDULE_UNSCHEDULED_VISIT))
```
Can the user make use of these in the form if they are only getting added at build time?

proteusvacuum replied:
they can, but they have to type out the paths themselves (i.e. they don't show up as hidden variables)

snopoke commented:
Didn't do a thorough review but generally looks good. Nothing blocking merge form me.

proteusvacuum committed:
make `pre_delete_hook` public

proteusvacuum committed:
Add logging error to pre_delete and pre_move hooks

proteusvacuum committed:
applyGlobalEventHandlers

proteusvacuum committed:
Add comment js comment

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/visit-scheduler.js:
```diff
@@ -121,24 +297,22 @@ var VisitScheduler = (function () {
 
+            self.relevancy = ScheduleRelevancy.wrap(data.schedule);
+            var xmlRe = /\s+|<+|>+|&+|"+|'+/g;
+            self.schedule_form_id = ko.observable(data.schedule_form_id).snakeCase(xmlRe);
```
what's this doing?

proteusvacuum replied:
since this gets used as an xml tag name, this removes characters that can't be used in a tag name (<,>,&, ", '), and replaces them with _

orangejenny replied:
nm, found the `snakeCase` definition. curious why you're passing a custom regex rather than stripping all non-word characters.

orangejenny replied:
...hadn't seen your reply when I commented again. Makes sense.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/visit-scheduler.js:
```diff
@@ -196,6 +379,40 @@ var VisitScheduler = (function () {
+                //figure out its new position
+                var position = ko.utils.arrayIndexOf(ui.item.parent().children(), ui.item[0]);
+                //remove the item and add it back in the right spot
+                if (position >= 0) {
```
why only `position >= 0`?

proteusvacuum replied:
This was actually me script-kiddying from the internetz. I'll add a link back to where I found this, and look over it more closely.

proteusvacuum replied:
I think this is because `arrayIndexOf` returns -1 if the item isn't found.

orangejenny replied:
oh, i totally read that as `position > 0` (even when copying & pasting), sorry for the derail

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/visit-scheduler.js:
```diff
@@ -196,6 +379,40 @@ var VisitScheduler = (function () {
+        if (valueAccessor()) {
+            setTimeout(function() {
+                $(element).focus().select();
+            }, 0); //new tasks are not in DOM yet
```
`_.defer` instead of `setTimeout` might make the intent clearer

orangejenny commented on corehq/apps/app_manager/templates/app_manager/form_view_base.html:
```diff
@@ -57,6 +57,15 @@
+            margin-left: 50px;
+        }
+        .nested-fieldset legend{
+            font-size: 1rem;
```
I didn't know about `rem`. Interesting.

proteusvacuum replied:
Get ready, Bootstrap 4 is coming...

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/enable_schedule.html:
```diff
@@ -1,29 +1,83 @@
+                <tbody data-bind="template: { name: 'phase-template', foreach: phases},
+                                  sortableList: phases"></tbody>
+            </table>
+            <button class="btn" data-bind="click: addPhase">{% trans "Add Phase" %}</button>
```
minor: how about a plus icon?

proteusvacuum replied:
:heavy_plus_sign: :+1:

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/enable_schedule.html:
```diff
@@ -1,29 +1,83 @@
+                              visible: forms().length == 0
+                              "></i>
+                <i style="cursor: pointer;"
+                   class="icon-question-sign icon-blue"
```
Can this use the more standard `hq-help-template`?

proteusvacuum replied:
I tried doing that, but it wasn't playing nicely with being able to drag and drop the table rows.

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/form_tab_visit_scheduler.html:
```diff
@@ -3,98 +3,202 @@
+                            </span>
+                            <span class="hq-help-template"
+                                  data-title="{% trans 'Schedule Phase' %}"
+                                  data-content="{% trans 'The phase this form is part of. Make phases on the module settings page' %}">
```
Could "module settings page" be a link?

orangejenny commented on corehq/apps/app_manager/tests/data/suite/schedule-entry.xml:
```diff
@@ -6,11 +6,20 @@
-            <next_39f0ea function="if(current_schedule_phase = 3 and dob != '' and (instance('schedule:m1:f2')/schedule/@expires = '' or today() &lt; (date(dob) + instance('schedule:m1:f2')/schedule/@expires)), date(dob) + int(instance('schedule:m1:f2')/schedule/visit[@id &gt; last_visit_number_39f0ea][@late_window = '' or today() &lt;= (date(dob) + int(@due) + int(@late_window))][1]/@due), 0)"/>
-            <next_due function="min(next_a1e369,next_a983e9,next_39f0ea)"/>
-            <is_late function="next_due &lt; today()"/>
+            <anchor_हिन्दी function="edd"/>
```
`<anchor_हिन्दी function="edd"/>` is legit XML? that's crazy.

proteusvacuum replied:
I actually need to verify this... I don't think my regex is tight enough to fit the spec (which does allow a small subset of unicode characters in tag names). Also, I just tested that example on the mobile app, and it crashes.

proteusvacuum replied:
but yes, apparently that is well formed XML. You can use this swanky JS feature I just discovered: `DOMParser()`
```js
var parser = new DOMParser()
var parsed = parser.parseFromString("<anchor_हिन्दी function='edd'/>", "application/xml")
parsed.activeElement.tagName === "parseerror" ? "error" : "good!"
```

orangejenny replied:
nice

orangejenny commented on corehq/apps/app_manager/views.py:
```diff
@@ -2083,11 +2117,53 @@ def should_edit(attribute):
+        module.update_schedule_phases(all_anchors)
+        module.has_schedule = enabled
+    except ScheduleError as e:
+        return HttpResponseBadRequest(unicode(e))
```
Can this instead redirect to a friendlier page with an error, a la https://github.com/dimagi/commcare-hq/pull/7790 ?

proteusvacuum replied:
It could... although it seems as though all of the other errors in app_manager are handled this way IIRC (with a javascript alert), which I hate, although it does reduce the number of page loads, and maintains consistency.

orangejenny replied:
If it does a js alert I'm more okay with it. JS alert isn't ideal, but unstyled error with no escape route is a lot worse.

proteusvacuum replied:
Yeah, it does a JS alert with the error message, and changes the save button text to "Try Again"

orangejenny commented:
This looks great. Nothing blocking.

proteusvacuum commented:
awesome, thanks for reading all that!. I have a few blocking QA things outstanding, but almost there. Also incorporating your and Simon's feedback.
#### QA outstanding:
- [ ] Unscheduled visits relevant condition not triggering correctly
- [ ] Add validation to ensure that at least one visit is scheduled. 

Working on fixes for those now, when that's done I think we'll be g2m.

proteusvacuum committed:
Unscheduled visits

If you are repeating the same visit, it is unscheduled.

proteusvacuum committed:
Use `session_case_id` instead of action to get case

proteusvacuum committed:
Dont steal code without referencing it

proteusvacuum committed:
Plus icon on button

proteusvacuum committed:
Add link to module settings page

proteusvacuum committed:
Fixup: add unicode to coalesce string

proteusvacuum committed:
Validate that there is at least one visit

proteusvacuum committed:
Merge branch 'master' into icds/scheduler

proteusvacuum committed:
Allow delecting of schedule if there are errors

proteusvacuum commented:
OK. G2M when :white_check_mark:

orangejenny commented:
looked over recent commits, agree is good to merge after tests

snopoke commented:
:tada:  nice work @proteusvacuum

czue commented:
:boom:

proteusvacuum commented:
yay!
